pageserver/src/tenant/storage_layer.rs

```diff
@@ -204,23 +204,52 @@ impl Default for ValuesReconstructState {
     }
 }
 
-/// Description of layer to be read - the layer map can turn
-/// this description into the actual layer.
-#[derive(PartialEq, Eq, Hash, Debug, Clone)]
-pub(crate) enum ReadableLayerDesc {
-    Persistent {
-        desc: PersistentLayerDesc,
-        lsn_range: Range<Lsn>,
-    },
-    InMemory {
-        handle: InMemoryLayerHandle,
-        lsn_ceil: Lsn,
-    },
+/// A key that uniquely identifies a layer in a timeline
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub(crate) enum LayerKey {
+    PersitentLayerKey(PersistentLayerKey),
+    InMemoryLayerKey(InMemoryLayerKey),
 }
 
-/// Wraper for 'ReadableLayerDesc' sorted by Lsn
+/// Layer wrapper for the read path. Note that it is valid
+/// to use these layers even after external operations have
+/// been performed on them (compaction, freeze, etc.).
 #[derive(Debug)]
-struct ReadableLayerDescOrdered(ReadableLayerDesc);
+pub(crate) enum ReadableLayer {
+    PersistentLayer(Layer),
+    InMemoryLayer(Arc<InMemoryLayer>),
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ReadDesc {
+    layer_key: LayerKey,
+    lsn_range: Range<Lsn>,
+}
+
+impl Ord for ReadDesc {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let ord = self.lsn_range.end.cmp(&other.lsn_range.end);
+        if ord == std::cmp::Ordering::Equal {
+            self.lsn_range.start.cmp(&other.lsn_range.start).reverse()
+        } else {
+            ord
+        }
```

Comment:
nit

```suggestion
        self.lsn_range.end.cmp(&other.lsn_range.end)
            .then_with(|| self.lsn_range.start.cmp(&other.lsn_range.start).reverse())
```

https://doc.rust-lang.org/stable/std/cmp/enum.Ordering.html#method.then_with is nice. Could be `then` is enough, unsure if clippy nags about this.

Reply:
Not sure I like this. To me it's harder to read, but maybe it's just the untrained rust eye.

Reply:
I guess it starts to look better after doing the early exiting for few more fields. Ok to resolve.